src/user-event/type/type.ts

```diff
@@ -1,20 +1,129 @@
 import { ReactTestInstance } from 'react-test-renderer';
-import { UserEventInstance } from '../setup';
-import { dispatchHostEvent, wait } from '../utils';
+import { getConfig } from '../../config';
 import { EventBuilder } from '../event-builder';
+import { ErrorWithStack } from '../../helpers/errors';
+import { UserEventInstance } from '../setup';
+import {
+  dispatchHostEvent,
+  wait,
+  getTextRange,
+  getTextContentSize,
+} from '../utils';
+
+import { parseKeys } from './parseKeys';
+
+export interface TypeOptions {
+  skipPress?: boolean;
+  submitEditing?: boolean;
+}
 
 export async function type(
   this: UserEventInstance,
   element: ReactTestInstance,
-  text: string
-) {
-  // TODO provide real implementation
+  text: string,
+  options?: TypeOptions
+): Promise<void> {
+  if (element.type !== getConfig().hostComponentNames?.textInput) {
+    throw new ErrorWithStack(
+      `type() works only with "TextInput" elements. Passed element has type "${element.type}".`,
+      type
+    );
+  }
+
+  const keys = parseKeys(text);
+
+  if (options?.skipPress !== true) {
+    await wait(this.config);
+    dispatchHostEvent(element, 'pressIn', EventBuilder.Common.touch());
+  }
+
   await wait(this.config);
```

Comment:
why wait here? The wait accounts for delay between user actions like typing right? So it should only occur between keys being typed no?

Reply:
In real life there is a slight delay between press in and press out events and focus is connected to press out

Reply:
press out is called after press ends and a minimum of 130ms have ellapsed after the start of the press, so it would make more sense I think to wait 130ms rather than waiting for the delay which accounts for user interactions

Reply:
I am wondering how realistic we want to be for events timing. There is a spectrum of possibilities:
* on the most realistic end we would try to check the timing of various events use it in respective functions
* on the other end we could follow RTL's UE approach and use delay of 0 ms by default (= next event queue iteration).

For some time-related event like `longPress` obviously we need to introduce the 500ms or so waiting, but for regular `press` and typing events we could go either of the above ways.

Pros and cons of realistic approach:
+ realistic timing
- user will need to use fake timers for all tests or have long running tests
- variable timing of different events, not consistent with RTL's UE `delay` option API
- need to conduct experiments to get actual timing

Pros and cons of lightweight approach:
+ simple implementation
+ consistent with RTL's UE `delay` option API
+ can run most of the tests without fake timers
+ simulation good enough for most cases (next event loop iteration)
- artificial (0ms or fixed) timing delay between events

If user is to take benefit of the realistic timing, e.g. for touch interactions calculation, we should probably restore non-zero `timestamp` for press events, as this might be useful in such cases.

So far the lightweight / RTL's UE approach seems to offer better tradeoffs, but I might have the whole picture of the situation.

@pierrezimmermannbam @MattAgn @AugustinLF  What are your thoughts on the event timing subject, is there any aspect I've omitted or misunderstood.

Reply:
For every timing that would represent the time for a user to do an action, e.g. between typing two characters or to press a button, I would agree that sticking to the RTL's approach looks better. I think that how much time is spent between two actions will very rarely matter and since it can vary a lot from a user to another the code should not to rely on that. Also the users currently still have the possibility to change that delay if they want to.

Then for some other situations, it can be interesting to have a more accurate timing. For press events on Pressables for instance, since they're not mocked by RN preset we don't really have a choice, we need to wait 500ms to trigger a long press and the pressOut only happens after a minimum 130ms after the press starts so we also have to wait for that. For press on Text and TextInput, we don't really have to wait but if the users specifies a press duration (which defaults to 0) then we should wait for that time in any case. Technically we don't have to wait before the press out because Text and Textinputs are currently mocked but since it's a RN rule that pressOut will be called after a minimum of 130ms I would still wait, even though it will make tests without fake times slower 

The event builder for press events does not use a zero timestamp and maybe we should reconsider using a zero timestamp for any of the events because the only motivation to do so if i remember correctly was to make test writing easier

Reply:
Thanks for sharing your perspective. Do you have more info on that min 130 ms wait time? Is that for RN docs, source code or experiments? I would like to get better understanding around this as it seems important in our context.

Reply:
This is from source code of [Pressability](https://github.dev/facebook/react-native/blob/6d24ee13a47a58ff1b19c5a51f39f52c9a4a8a28/packages/react-native/Libraries/Pressability/Pressability.js#L788-L789). I came across that when building press. I realized that onPressOut was not immediately called after calling onResponderRelease. That's why in my [implementation](https://github.com/callstack/react-native-testing-library/blob/dc7452284f352f2b366011610467d7f79d3365fa/src/user-event/press/press.ts#L104) I'm waiting for the press out when the press duration is bellow 130ms